research/gaia/pegasus/network/src/manager.rs

```diff
@@ -62,7 +63,34 @@ impl ServerManager {
 }
 
 impl ServerDetect for Vec<Server> {
-    fn fetch(&mut self) -> &[Server] {
-        self.as_slice()
+    fn fetch(&mut self) -> Vec<Server> {
+        self.clone()
+    }
+}
+
+pub struct SimpleServerDetector {
+    peers_mutex: Mutex<Vec<Server>>,
+}
+
+impl SimpleServerDetector {
+    pub fn new() -> Self {
+        SimpleServerDetector {
+            peers_mutex: Mutex::new(vec![]),
+        }
+    }
+
+    pub fn update_peer_view(&self, peer_view: Vec<(u64, SocketAddr)>) {
+        let new_peers = peer_view.into_iter()
+            .map(|(id, addr)| Server { id, addr, })
+            .collect::<Vec<Server>>();
+        let mut peers =  self.peers_mutex.lock().unwrap();
+        *peers = new_peers;
+    }
+}
+
+impl ServerDetect for Arc<SimpleServerDetector> {
```

Comment:
Better implement for `SimpleServerDetector` then for `Arc<SimpleServerDetector>`